internal/linker/linker.go

```diff
@@ -0,0 +1,263 @@
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if file.IsDelete || file.IsRename {
+				panic("delete and rename patch not supported")
+			}
+
+			patches[file.OldName] = string(patchBytes)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return "", nil, err
+	}
+	return base64.RawStdEncoding.EncodeToString(versionHash.Sum(nil)), patches, nil
+}
+
+func applyPatch(workingDirectory, patch string) error {
+	cmd := exec.Command("git", "-C", workingDirectory, "apply")
+	cmd.Stdin = strings.NewReader(patch)
+	return cmd.Run()
+}
+
+func copyFile(src, target string) error {
+	targetDir := filepath.Dir(target)
+	if err := os.MkdirAll(targetDir, os.ModePerm); err != nil {
+		return err
+	}
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	targetFile, err := os.Create(target)
+	if err != nil {
+		return err
+	}
+	defer targetFile.Close()
+	_, err = io.Copy(targetFile, srcFile)
+	return err
+}
+
+func existsFile(path string) bool {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+	return !stat.IsDir()
+}
+
+func applyPatches(srcDirectory, workingDirectory string) (map[string]string, error) {
+	mod := make(map[string]string)
+	for fileName, patch := range linkerPatches {
+		oldPath := filepath.Join(srcDirectory, fileName)
+		newPath := filepath.Join(workingDirectory, fileName)
+		mod[oldPath] = newPath
+
+		if err := copyFile(oldPath, newPath); err != nil {
+			return nil, err
+		}
+
+		if err := applyPatch(workingDirectory, patch); err != nil {
+			return nil, fmt.Errorf("apply patch for %s failed: %v", fileName, err)
+		}
+	}
+	return mod, nil
+}
+
+func cachePath(goExe string) (string, error) {
+	var cacheDir string
+	if val, ok := os.LookupEnv(garbleCacheDir); ok {
+		cacheDir = val
+	} else {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			panic(fmt.Errorf("cannot retreive user cache directory: %v", err))
+		}
+		cacheDir = userCacheDir
+	}
+
+	cacheDir = filepath.Join(cacheDir, cacheDirName)
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return "", err
+	}
+
+	return filepath.Join(cacheDir, "link"+goExe), nil
+}
+
+func getCurrentVersion(goVersion string) string {
+	return linkerPatchesVer + " " + goVersion
+}
+
+func checkVersion(linkerPath, goVersion string) (bool, error) {
+	versionPath := linkerPath + versionExt
+	version, err := os.ReadFile(versionPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	return string(version) == getCurrentVersion(goVersion), nil
+}
+
+func writeVersion(linkerPath, goVersion string) error {
+	versionPath := linkerPath + versionExt
+	return os.WriteFile(versionPath, []byte(getCurrentVersion(goVersion)), os.ModePerm)
+}
+
+type overlayFile struct {
+	Replace map[string]string
+}
+
+func compileLinker(workingDirectory string, overlay map[string]string, outputLinkPath string) error {
+	file, err := json.Marshal(&overlayFile{Replace: overlay})
```

Comment:
note that you don't need to declare a type for just one use:

```
file, err := json.Marshal(&struct{Replace map[string]string}{Replace: overlay})
```

Reply:
you can even omit `Replace:` on the right side I think, since it's obvious what the field is - being in the same line and the only field.

```
file, err := json.Marshal(&struct{Replace map[string]string}{overlay})
```